enhancements/etcd/scaling-etcd-with-raft-learners.md

```diff
@@ -0,0 +1,343 @@
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in
+  [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Over time as clusters live longer and workloads grow the ability to scale the control-plane and replace failed nodes
+becomes a critical part of the admins maintenance overhead. Today the `cluster-etcd-operator` manages scaling up of
+the etcd cluster. To provide the foundation for initiatives such as scale down and vertical control plane scaling[1].
+The `cluster-etcd-operator` must ensure proper safety mechanisms exist to adjust membership of the etcd cluster.
+
+Introduced in etcd v3.4 the raft learner[2] provides mitigations which reduces quorum and
+stability issues during scaling. A learner is essentially an etcd member which is non voting thus can not impact 
+quorum but like other members will receive log replications from the leader.
+
+This enhancement proposes:
+- Replacing the default scale up performed by the cluster-etcd-operator to use raft learners.
+- Deprecation and removal of the current `discover-etcd-initial-cluster`[3] command and replacing its
+functionality with the existing `etcdenvvar` controller and subcommands of the `cluster-etcd-operator`. By
+eliminating this code it also removes the teams largest carry against upstream etcd.
+- Ensuring that scale up only happens on control-plane nodes which have a quorum-guard pod monitoring health.
+- Adding a flag to etcd which allows configuration of maximum learners in cluster (--max-learners). Today the max is 1.
+
+POC: Functional proof of concept: https://github.com/openshift/cluster-etcd-operator/pull/682
+
+[1] https://issues.redhat.com/browse/OCPPLAN-5712
+
+[2] https://etcd.io/docs/v3.4/learning/design-learner/
+
+[3] https://github.com/openshift/etcd/blob/openshift-4.10/openshift-tools/pkg/discover-etcd-initial-cluster/initial-cluster.go
+
+## Motivation
+
+Great demand exists to provide flexible automated scaling of the control-plane in the same way the worker
+nodes do today.
+
+### Goals
+
+- provide safe scale up of etcd cluster using raft learners.
+- add additional membership level details such as member ID to the init container validation process for etcd.
+- add observability to allow admin to clearly diagnose scaling failure.
+- reduce divergence from upstream etcd by removing `discover-etcd-initial-cluster`.
+
+### Non-Goals
+
+- implement scale down logic for the `cluster-etcd-operator`
+
+## Proposal
+
+### User Stories
+
+1. As a cluster-admin I want to scale up etcd without fear of quorum loss.
+
+2. As a cluster-admin I want to be able to replace failed control-place nodes without a deep understanding of etcd.
+
+### Critical Alerts
+
+We will add the following critical alerts:
+
+- alert about learner member which has not been promoted or started > 30s
+- alert if the number of etcd members is not equal to the number of quorum-guard pods > 30s
```

Comment:
Is there a defined term (perhaps we can create one) for talking about all members vs only those which are voting members?

Does etcd quorum guard need to cover the learner? Or does that only need to to cover the member once it is promotable?

Reply:
Cool, that makes sense, might be worth adding that context into the enhancement as it wasn't obvious to me why learners still need quorum guard